lib/ditty/controllers/main.rb

```diff
@@ -145,21 +145,54 @@ def redirect_path
       if env['omniauth.auth']
         # Successful Login
         user = User.find(email: env['omniauth.auth']['info']['email'])
-        #Check Identity for Password Expiry
+        # Check Identity for Password Expiry
         password_expiry_check(user)
         self.current_user = user
-        broadcast(:user_login, target: self, details: "IP: #{request.ip}")
-        halt 200 if request.xhr?
-        flash[:success] = 'Logged In'
-        redirect redirect_path
+        if multi_factor_authentication?
+          identity = Identity.find(user_id: user[:id])
+          pin = mfa_pin(identity)
+          flash[:info] = 'An email was sent with your OTP'
+          Ditty::Services::Email.deliver(
+            :otp,
+            identity[:username],
+            locals: { identity: identity, otp: pin, request: request }
+          )
+          mfa_url = "#{settings.map_path}/auth/identity/mfa"
+          redirect mfa_url
+        else
+          successful_login
+        end
       else
-        # Failed Login
-        broadcast(:identity_failed_login, target: self, details: "IP: #{request.ip}")
-        flash[:warning] = 'Invalid credentials. Please try again.'
+        failed_login
+      end
+    end
+
+    get '/auth/identity/mfa' do
+      if current_user&.is_a? Ditty::User
+        identity = Identity.find(user_id: current_user[:id])
```

Comment:
Isn't there a `current_user_id` helper method? Not a trainsmash, but easier to read.